server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java

```diff
@@ -682,9 +682,16 @@ public RowSignature buildDataSourceRowSignature(final String dataSource)
           RowSignature rowSignature = optionalSchema.get().getSchemaPayload().getRowSignature();
           mergeRowSignature(columnTypes, rowSignature);
         } else {
-          // mark it for refresh, however, this case shouldn't arise by design
-          markSegmentAsNeedRefresh(segmentId);
           log.debug("SchemaMetadata for segmentId [%s] is absent.", segmentId);
+
+          ImmutableDruidDataSource druidDataSource =
+              sqlSegmentsMetadataManager.getImmutableDataSourceWithUsedSegments(segmentId.getDataSource());
+
+          if (druidDataSource != null && druidDataSource.getSegment(segmentId) != null) {
```

Comment:
Can we remove the segment if the segment is unused when we receive a callback from the inventory view.?

Reply:
Whenever `SqlSegmentsMetadataManager` polls the database and returns the latest list of used segments, we should check that list and remove any segments from cache which are not in that list.

This removal should not be triggered based on any callback from server inventory view. Datasource should be the source of truth on whether a segment is used or not.

Also, I was under the impression that this feature is already included. Can you confirm, @findingrish ?

Reply:
>Whenever SqlSegmentsMetadataManager polls the database and returns the latest list of used segments, we should check that list and remove any segments from cache which are not in that list.

This doesn't happen currently. Callbacks from the inventory view is the source of truth for the metadata cache. 
The reason being that segments in the metadata cache are representative of the available segments loaded on various data servers in the cluster.

Reply:
> Callbacks from the inventory view is the source of truth for the metadata cache.

I am not sure I understand. Isn't the schema coming from the metadata store?

Reply:
>I am not sure I understand. Isn't the schema coming from the metadata store?
That is partly correct, we still rely on metadata query to fetch schema for segments published from MSQ job. 
The datasource schema is still computed from the segments which are `available`.

Reply:
> This removal should not be triggered based on any callback from server inventory view. Metadata store should be the source of truth on whether a segment is used or not.

Another option is to register segment removal callback in various places that a segment is marked as unused. But that might be difficult to maintain.

Reply:
> Another option is to register segment removal callback in various places that a segment is marked as unused. But that might be difficult to maintain.

No, this is not right.

> That is partly correct, we still rely on metadata query to fetch schema for segments published from MSQ job.
The datasource schema is still computed from the segments which are available.

But these schemas would be populated back into the DB, right?

No service is going to request the coordinator for the schema of an unused segment.
And if it does and the coordinator _knows_ that the segment is unused, it can just return an error response saying that it doesn't know about this segment.

Reply:
> No service is going to request the coordinator for the schema of an unused segment.
And if it does and the coordinator knows that the segment is unused, it can just return an error response saying that it doesn't know about this segment.

That is completely true. Are you suggesting that we build datasource schema from the segment metadata and schema polled from the database?

Reply:
Yes, my understanding is that we should be doing the following:

- Every minute (or less), coordinator polls the metadata store to get the list of used segments and their schemas
- Coordinator populates its cache with this information
- Coordinator gets rid of any segment not found in the above list
- If any used segment that the Coordinator received in the latest poll does not have the schema populated, we fire a metadata query to get that schema and populate it back into the DB.
- Subsequent polls will get schema for that segment also.
- If any service requests schema for a used segment, coordinator sends back an OK response
- If any service requests schema for an unused segment, coordinator sends back a NOT FOUND response.

Let me know if this aligns with what we need and what we currently have.

---
Traditionally, coordinator has only been aware of "used" segments only (both overshadowed and non-overshadowed). We should continue to maintain that state.

Coordinator does also maintain an inventory view which might show unused segments being present on some server. But in the immediately next run, the Coordinator unloads the unused segments.

Reply:
I don't think we can base the schema building logic entirely on segments polled from the database since realtime segment schema isn't present in the metadata database. 

Also, in the present mechanism datasource schema is rebuilt only on segment addition/removal. In the proposed approach the datasource schema would be rebuilt after each poll leading to overhead?

Reply:
>Can we remove the segment if the segment is unused when we receive a callback from the inventory view.?

@cryptoe I think there could be a scenario where this would still fail. 

* Segment `s1` got added to `foo` and the cache tries to refresh it. 
* The cache is building the schema for `foo` and in parallel `s2` was marked unused. 
* The caches marks `s2` to be refreshed in the next cycle. 
* Lets assume that the segment removal callback didn't make it to the Coordinator until the next refresh cycle got triggered. 
* `s2` an unused segment will be refreshed in that cycle.

Reply:
> I don't think we can base the schema building logic entirely on segments polled from the database since realtime segment schema isn't present in the metadata database.

That's a fair point. How does the coordinator currently distinguish between a realtime segment and a published segment?

> Also, in the present mechanism datasource schema is rebuilt only on segment addition/removal. In the proposed approach the datasource schema would be rebuilt after each poll leading to overhead?

Don't you already do that? If not, how is the schema polled from the DB used?

Reply:
>How does the coordinator currently distinguish between a realtime segment and a published segment?

Both segments are added to the timeline, the difference is in refreshing the segment schema. 
Traditionally, a realtime segment would be refreshed every minute by issuing metadata query to tasks serving the segment whereas a published segment would be refreshed only once after the segment has been added. 

In the current scheme, schema for published segment is obtained from the database whereas realtime segment schema is periodically pushed to the Coordinator from tasks. 

> Don't we already do that? If not, how is the schema polled from the DB used?

No we don't do this as of now. The segment polled from the database is cached and used on-demand ie. when a segment is added or datasource schema is being rebuilt.